### DISCUSSION TITLE
Handle multiple photos for bot

### DISCUSSION THREAD

**Discussion by AguChaves92** [telegraf/telegraf#1806] [Q&A]
I have this code inside a wizard scene: 
`async (ctx) => {
    const photos = ctx.message.photo;

    const filteredPhotos = photos.filter((photo) => photo.width === 90);
    if (ctx.wizard.state.product.photos === undefined) {
      ctx.wizard.state.product.photos = [];
    }
    ctx.wizard.state.product.photos.push(...filteredPhotos);

    const media = [];

    media.push({
      type: "photo",
      media: ctx.wizard.state.product.photos[0].file_id,
      caption: "Hello",
    });

    // Add remaining photos without a caption
    for (let i = 1; i < ctx.wizard.state.product.photos.length; i++) {
      media.push({
        type: "photo",
        media: ctx.wizard.state.product.photos[i].file_id,
      });
    }

    await ctx.telegram.sendMediaGroup("@mychannel", media);
    ctx.reply(`Tu producto ha sido registrado   
    `);
    return ctx.scene.leave();
  }`
  
 the thing is that when the user sends more than one picture, the reply to the group is sent for every picture the user sent. Is there a way to capture all pictures and only after receiving them all executing the reply part?

**Comment by MKRhere** [comment]
There isn't a way to know when all photos have arrived, or to know how many photos the user sent. Telegraf will just receive the photos one by one. If they were part of an album, every photo will have a matching `ctx.message.media_group_id`. You'll need other creative solutions to this problem.

**Comment by neoquijo** [comment]
know it's bit late but maybe will be useful for sommeone:



```
const accumulators: { [key: string]: { photos: any[], timeout: NodeJS.Timeout | null } } = {};

export const uploadUserPhotos = async (message: any): Promise<void> => {

    const images: Array<string> = [];
    const mediaGroupId = message.media_group_id;

    // Block executed if user sends multiple photos (media_group_id exists)
    if (mediaGroupId) {
        if (!accumulators[mediaGroupId]) {
            accumulators[mediaGroupId] = { photos: [], timeout: null };
        }

        // Add the current message to the accumulator
        accumulators[mediaGroupId].photos.push(message);

        // Clear any existing timeout to wait for all images in the group
        if (accumulators[mediaGroupId].timeout) {
            clearTimeout(accumulators[mediaGroupId].timeout);
        }

        // Set a new timeout to process all images in the group
        accumulators[mediaGroupId].timeout = setTimeout(async () => {
            const photos = accumulators[mediaGroupId].photos;

            for (const msg of photos) {
                // Process each photo in the media group
                const photo = msg.photo[msg.photo.length - 1];
                const fileId = photo.file_id;

                // Add logic to upload photo and add the result to images
                // Example: images.push(await uploadPhoto(fileId));
            }

            // Ensure there are images to return
            if (images.length === 0) throw new Error('Something went wrong');
            // Example callback
            // cb(images);

            // Clean up the accumulator
            delete accumulators[mediaGroupId];
        }, 1500);

    } else {
        // Block executed if user sends a single photo
        const photo = message.photo[message.photo.length - 1];
        const fileId = photo.file_id;

        // Add logic to upload photo and add the result to images
        // Example: images.push(await uploadPhoto(fileId));

        // Ensure there is an image to return
        if (images.length === 0) throw new Error('Something went wrong');
        // Example callback
        // cb(images);
    }
};

// Function to upload a single photo (placeholder)
const uploadPhoto = async (fileId: string): Promise<string> => {
    // Logic to upload photo
    // Example: return (await axios.post('YOUR_URL', { file_id: fileId })).data.filePath;
    return 'photo_url'; // Placeholder return
};
```

**Comment by Nikita0x** [comment]
I dont like the solution with timeouts

I did it like so: 

If the user has not uploaded any photos yet (`userPhotos.length === 0`) - then just push it to the array where you store all user photos.
If `userPhotos.length !== 0` then append it at the end. 
In both cases, reply back to user with the counter of how many photos they have uploaded : 
```ts
await ctx.reply("Your photo received! ${userPhotos.length}/10")
```

And before that, make a check:
```ts
if (userPhotos.length >= 10) {
     await ctx.reply("You have reached the limit of photos.")
      return 
 }
 ```
 
 We will also have to introduce here a button for the user to confirm when they are done - but it removes the need for any timeouts, guarantees that we receive all the photos and provides a decent UX for the user
